Feat/dsm/custom search

wi-kshitij-nagvekar opened the pull request.

wi-kshitij-nagvekar committed:
feat(SearchInputWithAdd): introduce custom search input component with add functionality

- Added `SearchInputWithAdd` component that extends the standard `SearchInput` to allow users to add new items when no search results are found.
- Implemented filtering, item selection, and customizable callbacks for adding new items.
- Updated documentation to include usage examples and TypeScript types for the new component.
- Registered the component in the main index file and added relevant entries in the knowledge base.

wi-kshitij-nagvekar committed:
chore: update fonts and dependencies

- Added a global CSS reset to ensure consistent box model calculations across elements.
- Updated package.json to include the latest version of `@testing-library/user-event`.
- Removed the deprecated `src/fonts.css` file as it was no longer needed.

wi-kshitij-nagvekar committed:
feat(SearchInputWithAdd): add unit tests for SearchInputWithAdd component

- Introduced comprehensive tests for the SearchInputWithAdd component, covering rendering, search functionality, item selection, and add new item features.
- Implemented tests for controlled and uncontrolled modes, ensuring proper behavior in various scenarios.
- Enhanced test coverage for edge cases, including handling empty items and custom rendering functions.

wi-kshitij-nagvekar committed:
feat(App): integrate SearchInputWithAdd component for tag management

- Added SearchInputWithAdd component to the App, allowing users to search and add tags dynamically.
- Implemented state management for available and selected tags, enhancing user interaction.
- Included UI elements for displaying selected tags and providing feedback on tag addition.

wi-kshitij-nagvekar committed:
chore: update @testing-library/user-event to version 14.6.1 in package-lock.json

- Added @testing-library/user-event version 14.6.1 to package-lock.json, ensuring compatibility with peer dependencies and enhancing testing capabilities.

wi-kshitij-nagvekar committed:
feat(SearchInputWithAdd): implement new component with customizable add functionality

- Introduced `SearchInputWithAdd` component that enhances the existing `SearchInput` by allowing users to add new items when search results are empty.
- Added props for customizable behavior, including `addNewItemText`, `filterFn`, and `renderItem`.
- Implemented controlled and uncontrolled modes for managing search terms, along with an `onAddNewItem` callback.
- Updated documentation in the knowledge base for the new component.
- Added global CSS reset for consistent layout and updated Jest configuration to improve test coverage.

nogoodusername committed:
Merge branch 'main' into feat/dsm/custom-search

nogoodusername committed:
Merge branch 'main' into feat/dsm/custom-search

wi-paras-mehta commented on packages/dsm/src/components/search-input-with-add/search-input-with-add.tsx:
```diff
@@ -0,0 +1,307 @@
+
+  const hasResults = filteredItems.length > 0;
+  const shouldShowAddOption =
+    !hasResults && currentSearchTerm.trim() !== '' && showAddNewOption;
```
Current behaviour always shows the add button, we can question it though

nogoodusername replied:
Sure, will explore this

wi-paras-mehta commented on packages/dsm/src/components/search-input-with-add/search-input-with-add.tsx:
```diff
@@ -0,0 +1,307 @@
+   */
+  filterFn?: (
+    item: { id: string; title: string; [key: string]: unknown },
+    searchTerm: string,
```
I think we should keep the searchList pass as id, label and value , or again i need to search the value from the label in the list passed where i use the component

nogoodusername commented on packages/dsm/src/components/search-input-with-add/search-input-with-add.tsx:
```diff
@@ -0,0 +1,307 @@
+    return item.title.toLowerCase().includes(term.toLowerCase());
+  };
+
+  const activeFilterFn = filterFn ?? defaultFilterFn;
```
An empty filter function is what we would need to place here for online search implementation

wi-kshitij-nagvekar committed:
feat(SearchInputWithAdd): enhance add option behavior and update documentation

- Introduced `addOptionBehavior` prop to control when the "Add New" option is displayed: 'never', 'no-results-only' (default), or 'always'.
- Updated `SearchInputWithAdd` component to reflect the new behavior, allowing for more flexible user interactions.
- Enhanced documentation to include examples and clarify the usage of the new prop.
- Added tests to verify the functionality of the new add option behavior.
